lib/plug/crypto/message_encryptor.ex

```diff
@@ -9,34 +9,37 @@ defmodule Plug.Crypto.MessageEncryptor do
   This can be used in situations similar to the `Plug.Crypto.MessageVerifier`,
   but where you don't want users to be able to determine the value of the payload.
 
-  The current algorithm used is AES-GCM-128.
+  The current algorithm used is ChaCha20-and-Poly1305.
 
   ## Example
 
       iex> secret_key_base = "072d1e0157c008193fe48a670cce031faa4e..."
       ...> encrypted_cookie_salt = "encrypted cookie"
-      ...> encrypted_signed_cookie_salt = "signed encrypted cookie"
-      ...>
       ...> secret = KeyGenerator.generate(secret_key_base, encrypted_cookie_salt)
-      ...> sign_secret = KeyGenerator.generate(secret_key_base, encrypted_signed_cookie_salt)
       ...>
       ...> data = "José"
-      ...> encrypted = MessageEncryptor.encrypt(data, secret, sign_secret)
-      ...> MessageEncryptor.decrypt(encrypted, secret, sign_secret)
+      ...> encrypted = MessageEncryptor.encrypt(data, secret, "UNUSED")
+      ...> MessageEncryptor.decrypt(encrypted, secret, "UNUSED")
       {:ok, "José"}
 
   """
 
   @doc """
   Encrypts a message using authenticated encryption.
 
+  The `sign_secret` is currently only used on decryption
+  for backwards compatibility.
+
   A custom authentication message can be provided.
   It defaults to "A128GCM" for backwards compatibility.
   """
   def encrypt(message, aad \\ "A128GCM", secret, sign_secret)
-      when is_binary(message) and (is_binary(aad) or is_list(aad)) and byte_size(secret) > 0 and
+      when is_binary(message) and (is_binary(aad) or is_list(aad)) and
+             bit_size(secret) in [128, 192, 256] and
              is_binary(sign_secret) do
-    aes128_gcm_encrypt(message, aad, secret, sign_secret)
+    iv = :crypto.strong_rand_bytes(12)
+    {cipher_text, cipher_tag} = block_encrypt(:chacha20_poly1305, secret, iv, {aad, message})
+    encode_token("C20P1305", iv, cipher_text, cipher_tag)
```

Comment:
I don't know if message size is a concern, but if it is, there's no need for the `C20P1305` string. V1/V2 can be differentiated based on the number of `.` separate parts. If future versions are ambiguous, they can add some versioning prefix.

Reply:
Yeah, we are also base64 encoding it, which we don't need. I think I will make it smaller and stop encoding it at least.